### PR TITLE
Add methods to write IForgeRegistryEntry to ByteBuf

### PR DESCRIPTION
**Pull request by diesieben07**
This allows people to efficiently send IForgeRegistryEntry references over the network, without publicly exposing the ID.
Comments on the javadoc welcome.

### CODE REVIEW

**Comment by mezz** [comment]
I think it would be better to implement this in `FMLControlledNamespacedRegistry` and expose it as an interface on `IForgeRegistry`.

**Comment by diesieben07** [comment]
I thought about that, but it doesn't seem right to me for the registry to deal with the serialization to bytes. And also it fits in with everything else being handled through ByteBufUtils. But if you want i'll move it.

**Comment by mezz** [comment]
I think `checkRegistry` here is uglier than having the registry handle serialization itself.

**Comment by diesieben07** [comment]
Moved the methods to `IForgeRegistry`.

**Comment by Alpvax** [comment]
Handling this in the registry would be the easier way to go: `writeRegistryEntry(ByteBuf to, T entry)`, but for consistency it should be in ByteBufUtils, along with writing NBT and UUIDs.

File: src/main/java/net/minecraftforge/fml/common/registry/FMLControlledNamespacedRegistry.java [hunk 1]
```diff
@@ -935,4 +938,15 @@ public void remove()
         return new RegistryEvent.Register<I>(location, this);
     }
 
+    @Override
+    public void writeEntry(ByteBuf out, I entry)
```
Comment (Alpvax) [inline src/main/java/net/minecraftforge/fml/common/registry/FMLControlledNamespacedRegistry.java]: Should these parameters be marked `@Nonnull`?

**Comment by mezz** [comment]
`ByteBufUtils` is for varint helper methods and for handling non-forge classes without adding patches, I don't think there is an issue with consistency here.

File: src/main/java/net/minecraftforge/fml/common/registry/FMLControlledNamespacedRegistry.java (hunk 1 above)
Comment (mezz) [inline src/main/java/net/minecraftforge/fml/common/registry/FMLControlledNamespacedRegistry.java]: The nullability of Forge registries is ambiguous. Minecraft uses AIR as default values for its registries, but Forge's can return null or a default value.

File: src/main/java/net/minecraftforge/fml/common/registry/FMLControlledNamespacedRegistry.java [hunk 2]
```diff
@@ -935,4 +938,15 @@ public void remove()
...
     }
 
+    @Override
+    public void writeEntry(@Nonnull ByteBuf out, @Nonnull I entry)
+    {
+        ByteBufUtils.writeVarInt(out, getId(entry), 5);
+    }
+
+    @Override
+    public I readEntry(@Nonnull ByteBuf in)
```
Comment (mezz) [inline src/main/java/net/minecraftforge/fml/common/registry/FMLControlledNamespacedRegistry.java]: This should return `@Nullable` to match `FMLControlledNamespacedRegistry#getObjectById`

**Comment by Alpvax** [comment]
Are there any other objects that are not written either directly or via ByteBufUtils?
Is it worth adding a method to ByteBufUtils which just delegates to the registry?

**Comment by mezz** [comment]
`FMLMessage.EntitySpawnMessage#toBytes`
`FMLMessage.EntitySpawnMessage#fromBytes`
`Packet#readPacketData`
`Packet#writePacketData`

I don't think extra methods in `ByteBufUtils` is necessary, I don't think it's meant to be a universal serialization class.

**Comment by Alpvax** [comment]
Ok, looks good as-is then.

**Comment by LexManos** [comment]
This actually needs to be two var ints, one for the registry ID, the other for the entry in that registry.

**Comment by diesieben07** [comment]
I don't really see how that would make sense. What would `ForgeRegistries.BLOCKS.readEntry` return when it reads the registry id for `ForgeRegistries.ITEMS`? It can't return an item...

**Comment by LexManos** [comment]
Ah you put these on the registry objects, I'd rather not do that. I'd lock it down to the NBT helper/PacketHelper classes.
As for what it would return if it tried to read a entry for a invalid registry, it'd error.

**Comment by diesieben07** [comment]
God damn it @mezz 😄 
I will fix this up tomorrow.

**Comment by mezz** [comment]
This is for writing directly to packet buffer, using minimal data and a known ordering.
For NBT serialization it should store both registry ID and entry ID I agree, but not for this IMO.

**Comment by Alpvax** [comment]
I agree with @mezz.
Unless you are moving the methods to the registry entry, in which case , you should also be checking whether the entry is INBTSerializable and if so, (de)serialise it. Abstracting the whole process of syncing entries seems like overkill.

I cannot think of any case where you would be trying to deserialise an entry without knowing what type it is, so I don't think the registry id is necessary. If you want to serialise several types into 1 NBTTag, the modder can handle that themselves.

**Comment by diesieben07** [comment]
NBT is a different case altogether anyways, you should use the registry name there, not the ID.

**Comment by Alpvax** [comment]
Sorry, I got slightly thrown by Lex mentioning NBT.
Second point still applies however: When do you need the registry id? Unless forge is going to add an extendable SyncRegistryEntry packet, but I don't really see the point.

**Comment by diesieben07** [comment]
I completely agree.

**Comment by Alpvax** [comment]
Any idea when this will get pulled?

**Comment by liach** [comment]
Wait until it is assigned to lex and approved by mezz.

**Comment by diesieben07** [comment]
Can we get an official comment on whether it should still be moved back to the ByteBufUtils?

**Comment by diesieben07** [comment]
Another two weeks have passed.

> Can we get an official comment on whether it should still be moved back to the ByteBufUtils?

**Comment by LexManos** [comment]
Did I stutter?
Ad it to the packet buffer so that it's only for networking. 
Also has anyone looked into if the registry types are synced with IDs so you can compress it to two int.
Yes I know you said it *should* be a known format and known registry type. But I still prefer the idea of having the type id for better verification both internally and externally.

**Comment by diesieben07** [comment]
No, of course you did not stutter. But I assumed you would at least take _some_ consideration to the fact that mezz (your main PR review guy) and two other people disagreed with your opinion, hence I asked for re-confirmation. I guess that was not the case.

**Comment by diesieben07** [comment]
As for the registry ID: It does not seem to be synced, so sending the registry type via the byte buffer is not really viable.
This is the reason I asked again what to do...

**Comment by Alpvax** [comment]
If we are moving this back to the packet buffer is it worth adding (as Lex mentioned) a method to the NBT side of things which saves the ResourceLocations? Or do we just let the modders do that themselves?

**Comment by williewillus** [comment]
that's just a string, modders can do that themselves (as they probably are now)

**Comment by diesieben07** [comment]
Update:

- Methods moved back to `ByteBufUtils`.
- I had to introduce a new method `getRegistryName` in `IForgeRegistry` to obtain the name of the registry. This should be fine though, no mod should be implementing this manually.
- There is a slight source-incompatible (but binary compatible) change in `IForgeRegistryEntry` so that `getRegistryType` is actually usable with G`ameRegistry.findRegistry` without unsafe casts.
- To mitigate the overhead of writing the registry name as a string to the stream (registries do not have integer IDs currently) there are methods to write/read a list of entries which only write the registry name once.

Should I add a test mod?

File: src/main/java/net/minecraftforge/fml/common/network/ByteBufUtils.java [hunk 3]
```diff
@@ -219,6 +231,111 @@ public static NBTTagCompound readTag(ByteBuf from)
...
+                {
+                    entry = it.next();
+                }
+                else
+                {
+                    break;
+                }
+            }
+            while (true);
+        }
```
Comment (liach) [inline src/main/java/net/minecraftforge/fml/common/network/ByteBufUtils.java]:
```java
        if (it.hasNext())
        {
            T entry = it.next();
            FMLControlledNamespacedRegistry<T> registry = (FMLControlledNamespacedRegistry<T>) GameRegistry.findRegistry(entry.getRegistryType());
            writeUTF8String(out, registry.getRegistryName().toString());
            writeVarInt(out, registry.getId(entry), 5);
            while (it.hasNext())
            {
                writeVarInt(out, registry.getId(it.next()), 5);
            }
        }
```

**Review by liach — APPROVED** [review]

**Comment by liach** [comment]
Why is this pull ignored? This is really helpful.

File: src/main/java/net/minecraftforge/fml/common/registry/IForgeRegistryEntry.java [hunk 4]
```diff
@@ -59,7 +59,7 @@
     @Nullable
     ResourceLocation getRegistryName();
 
-    Class<? super V> getRegistryType();
```
Comment (LexManos) [inline src/main/java/net/minecraftforge/fml/common/registry/IForgeRegistryEntry.java:62]: Why are you changing this?

File: src/main/java/net/minecraftforge/fml/common/registry/IForgeRegistry.java [hunk 5]
```diff
@@ -40,6 +40,8 @@
 {
     public Class<V> getRegistrySuperType();
 
+    ResourceLocation getRegistryName();
+
```
Comment (LexManos) [inline src/main/java/net/minecraftforge/fml/common/registry/IForgeRegistry.java]: Rather not add this to the user facing API, the entire point is to hide things like this.

File: src/main/java/net/minecraftforge/fml/common/registry/FMLControlledNamespacedRegistry.java [hunk 6]
```diff
@@ -98,10 +99,11 @@
     @Nullable
     private final SubstitutionCallback<I> substitutionCallback;
 
-    FMLControlledNamespacedRegistry(@Nullable ResourceLocation defaultKey, int minIdValue, int maxIdValue, Class<I> type, BiMap<ResourceLocation, ? extends IForgeRegistry<?>> registries, @Nullable AddCallback<I> addCallback, @Nullable ClearCallback<I> clearCallback, @Nullable CreateCallback<I> createCallback, @Nullable SubstitutionCallback<I> substitutionCallback)
+    FMLControlledNamespacedRegistry(@Nullable ResourceLocation defaultKey, int minIdValue, int maxIdValue, Class<I> type, ResourceLocation registryName, BiMap<ResourceLocation, ? extends IForgeRegistry<?>> registries, @Nullable AddCallback<I> addCallback, @Nullable ClearCallback<I> clearCallback, @Nullable CreateCallback<I> createCallback, @Nullable SubstitutionCallback<I> substitutionCallback)
```
Comment (LexManos) [inline src/main/java/net/minecraftforge/fml/common/registry/FMLControlledNamespacedRegistry.java]: The implementation shouldn't need to know its name, the registry manager knows it.

File: src/main/java/net/minecraftforge/fml/common/network/ByteBufUtils.java (hunk 3 above)
Comment (LexManos) [reply src/main/java/net/minecraftforge/fml/common/network/ByteBufUtils.java]: even better, you could skip the iterator and just use get(0) and an enhanced for loop, bit this works fine.

File: src/test/java/net/minecraftforge/debug/ObjectHolderTest.java [hunk 7]
```diff
@@ -93,7 +93,7 @@ public ResourceLocation getRegistryName()
     }
 
     @Override
-    public Class<? super ICustomRegistryEntry> getRegistryType()
+    public Class<ICustomRegistryEntry> getRegistryType()
```
Comment (LexManos) [inline src/test/java/net/minecraftforge/debug/ObjectHolderTest.java:96]: again.. why?

**Comment by Alpvax** [comment]
> Why are you changing this?

> There is a slight source-incompatible (but binary compatible) change in IForgeRegistryEntry so that getRegistryType is actually usable with GameRegistry.findRegistry without unsafe casts.

File: src/main/java/net/minecraftforge/fml/common/network/ByteBufUtils.java (hunk 3 above)
Comment (diesieben07) [reply src/main/java/net/minecraftforge/fml/common/network/ByteBufUtils.java]: There is no `get(int)` on Collection.

File: src/main/java/net/minecraftforge/fml/common/registry/IForgeRegistry.java (hunk 5 above)
Comment (diesieben07) [reply src/main/java/net/minecraftforge/fml/common/registry/IForgeRegistry.java]: I can move it to FMLControlledNamespacedRegistry, but I thought registry names were public. Oh well.

**Comment by diesieben07** [comment]
Addressed the issues.

File: src/main/java/net/minecraftforge/fml/common/network/ByteBufUtils.java (hunk 3 above)
Comment (LexManos) [reply src/main/java/net/minecraftforge/fml/common/network/ByteBufUtils.java]: Well, atleast get rid of the horrid do while(true) as liach suggested, not only is it 11 lines longer then it needs to be, it's also a nasty looking piece of code. (Seriously if you ever have to do a while(true) you're *probably* doing it wrong.)

Comment (liach) [reply src/main/java/net/minecraftforge/fml/common/network/ByteBufUtils.java]: To be neater, we can just return when the iterator has no entries, less indentation.

**Comment by diesieben07** [comment]
Not sure why you dislike while(true) so much, but here you go.

**Comment by Alpvax** [comment]
> Not sure why you dislike while(true) so much, but here you go.

Just seems strange to have a while true and an if break, rather than having the condition itself in the while

**Comment by diesieben07** [comment]
> Just seems strange to have a while true and an if break, rather than having the condition itself in the while

It's the only way to avoid duplicated code. If you look at the code now you can see that the code to write the ID is duplicated.

**Comment by LexManos** [comment]
Whatever its good enough, better then while(true)

**Pull request merged**